src/ccs/ir/circuit.rs

```diff
@@ -0,0 +1,579 @@
+use std::{collections::HashMap, hash::Hash};
+
+use super::{assignments::*, CoeffsForProds};
+use crate::{
+    ccs::compiler::step_selector::{SelectorsForALLMatrix, SelectorsForALLSteps, StepSelector},
+    field::Field,
+    util::UUID,
+};
+
+pub type MatrixCoeffsAndOffset<F> = Vec<Vec<(CoeffsForProds<F>, usize)>>;
+pub type SelectorsOffsetAndCoeff<F> = Vec<Vec<(usize, F)>>;
+
+// ((step_idx, step_UUID), assignment_offset)
+pub type AssignmentOffsets = HashMap<(usize, u128), usize>;
+
+pub fn hadamard_product<F: Field + From<u64> + Hash>(vec1: &[F], vec2: &[F]) -> Vec<F> {
+    assert_eq!(vec1.len(), vec2.len());
+    vec1.iter()
+        .zip(vec2.iter())
+        .map(|(&v1, &v2)| v1 * v2)
+        .collect()
+}
+
+pub fn vec_add<F: Field + From<u64> + Hash>(vec: &[Vec<F>]) -> Vec<F> {
+    assert!(vec.len() > 1);
+    vec.iter().fold(vec![F::ZERO; vec[0].len()], |acc, vec| {
+        acc.iter().zip(vec.iter()).map(|(&a, &v)| a + v).collect()
+    })
+}
+
+// input vector z = [1, x, w]
+pub struct Z<F: Field + From<u64>> {
+    pub n: usize,
+    pub l: usize,
+    pub assignments: Vec<F>,
+    pub public_inputs: Vec<F>,
+}
+
+impl<F: Field + From<u64> + Hash> Z<F> {
+    pub fn new(n: usize, l: usize) -> Self {
+        assert!(n > l);
+        Self {
+            n,
+            l,
+            assignments: Default::default(),
+            public_inputs: Default::default(),
+        }
+    }
+
+    pub fn write(
```

Comment:
@10to4 there should be a way to create Z with an intitial value. If you are always calling `new` and then `write` they should be the same method

Reply:
Got it.